### PR TITLE
Use named constants in the ArbOS source code

### PR DESCRIPTION
**Pull request by edfelten**
* Add a list of pre-defined constants that are available in every Mini program
* Constants can also be defined in a Mini source code file, after the `use` statements, using syntax like `const my constant = 42;`
* At present, constants must be unsigned integers, but they could be extended in principle to support signed int, boolean, and None, and empty-tuple
* Allow use of a named constant anywhere an unsigned integer literal is allowed, using the syntax `const::name`
* Use constants throughout the ArbOS source code, for clarity

Some compromises were made to reduce implementation complexity. For example, the predefined constants are in the compiler's source code, rather than defined in Mini syntax and parsed by the compiler. Also, constants defined in one file can't be referenced in another file. And we require constants to be unsigned integers so they can easily be used in places where unsigned integers are required, such as declaring the sizes of fixed-size arrays.

### CODE REVIEW

File: arb_os/arbsys.mini [hunk 1]
```diff
@@ -437,52 +437,52 @@ impure func arbAddressTable_addressExists(topFrame: EvmCallFrame, calldata: Byte
...
+        evmOp_revert_knownCodePc(address(const::Address_ArbAddressTable), 160, 0, 0);
     }
     let offset = bytearray_get256(calldata, 36);
     let bytesLen = bytearray_get256(calldata, 68);
     if (offset >= bytesLen) {
-        evmOp_revert_knownCodePc(address(102), 161, 0, 0);  // buffer is too short
+        evmOp_revert_knownCodePc(address(const::Address_ArbAddressTable), 161, 0, 0);  // buffer is too short
     }
-    let stream = bytestream_new(bytearray_extract(calldata, 100+offset, bytesLen-offset));
+    let stream = bytestream_new(bytearray_extract(calldata, const::Address_ArbSys+offset, bytesLen-offset));
```
Comment (hkalodner) [inline arb_os/arbsys.mini:485]: This doesn't seem like a correct use of this constant

File: arb_os/arbsys.mini [hunk 2]
```diff
@@ -499,18 +499,18 @@ impure func arbAddressTable_decompress(topFrame: EvmCallFrame, calldata: ByteArr
...
-        evmOp_revert_knownCodePc(address(102), 163, 0, 0);
+        evmOp_revert_knownCodePc(address(const::Address_ArbAddressTable), 163, 0, 0);
     }
 }
 
 impure func arbAddressTable_compress(topFrame: EvmCallFrame, calldata: ByteArray) { // (address addr) -> bytes
 // compress addr, based on the current address table contents, and return the resulting buffer
     if (bytearray_size(calldata) != 36) {
-        evmOp_revert_knownCodePc(address(102), 100, 0, 0);
+        evmOp_revert_knownCodePc(address(const::Address_ArbAddressTable), const::Address_ArbSys, 0, 0);
```
Comment (hkalodner) [inline arb_os/arbsys.mini:513]: This seems like another incorrect use of `const::Address_ArbSys`

File: arb_os/evmCallStack.mini [hunk 3]
```diff
@@ -216,16 +216,33 @@ public impure func initEvmCallStack(
...
+                        const::TxResultCode_insufficientBalance,
+                        None<ByteArray>,
+                        None<EvmLogs>,
+                        Some(gasUsage)
+                    );
                 } else {
                     // unknown error type
-                    emitTxReceipt(request.incomingRequest, 255, None<ByteArray>, None<EvmLogs>, None<GasUsage>);
+                    emitTxReceipt(
+                        request.incomingRequest, const::TxResultCode_unknownFailure,
```
Comment (hkalodner) [inline arb_os/evmCallStack.mini:229]: The formatting here should probably split this line

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: arb_os/arbsys.mini (hunk 1 above)
Comment (edfelten) [reply arb_os/arbsys.mini:485]: Fixed

File: arb_os/arbsys.mini (hunk 2 above)
Comment (edfelten) [reply arb_os/arbsys.mini:513]: Fixed.

File: arb_os/evmCallStack.mini (hunk 3 above)
Comment (edfelten) [reply arb_os/evmCallStack.mini:229]: Fixed.

**Review by hkalodner — APPROVED** [review]
LGTM

**Pull request merged**